File: ADC/python/lib/waveshare_ADS1113/ADS1113.py
#!/usr/bin/python
# -*- coding:utf-8 -*-

# from . import I2C_config
import time
import I2C_config
import smbus


class ADS1113:
    def __init__(self,add ):
        self.i2c = I2C_config.I2C_config(add)       
        
    def Read_Word_Reg(self, Reg):
        data =  self.i2c.I2C_Read_Word(Reg)
        data = data>>8 | ((data<<8)&0xff00)
        return data
        
    def Write_Word_Reg(self, Reg, data):
        self.i2c.I2C_Send_Word (Reg, data)
        
    def ADS1113_Init(self):
        self.Write_Word_Reg(0x01,0x00E0); 
        #Continuous mode 
        #860 SPS

    def ADS1113_ReadData (self):
        return self.Read_Word_Reg(0x00)

    def ADS1113_GetValue(self):
        data = self.Read_Word_Reg(0x00)
        if((data>>15)>0):
            data = (((~data) & 0x7fff )* 2.048 / 0x7fff) * -1.0
        else :
            data = (((~data) & 0x7fff )* 2.048 / 0x7fff)
        return data
    
if __name__=="__main__":
    ADC = ADS1113(add = 0x48)
    ADC.ADS1113_Init()
    while(1):
        value = ADC.ADS1113_GetValue()
        print ("AD : %f\r\n"%value)
        time.sleep(1)
        
        

File: ADC/c/lib/ADS1113/ADS1113.c
#include "ADS1113.h"


UWORD ADS1113_Read_Reg(UBYTE Reg)
{
    UBYTE buf[2] = {Reg};
    I2C_Write_nByte(buf, 1);
    I2C_Read_nByte(buf,2);
    return buf[0]<<8 |  buf[1];
}

void ADS1113_Write_Reg(UBYTE Reg, UWORD data)
{
    UBYTE buf[3] = {Reg, (data >> 8) & 0xff, data & 0xff};
    // printf("fasong 0x%x 0x%x 0x%x\r\n",buf[0], buf[1], buf[2]);
    I2C_Write_nByte(buf, 3);
}

void ADS1113_Init(void)
{
    DEV_I2C_Init(ADS1113_Address);
    ADS1113_Write_Reg(0x01,0x00E0);
    // printf("ADS1113_Read_Reg = 0x%x \r\n",ADS1113_Read_Reg(0x01));
}


UWORD ADS1113_ReadData (void)
{ 
    UBYTE buf[2];
    buf[0] = 0x00; //00
    I2C_Write_nByte(buf, 1);
    I2C_Read_nByte(buf,2);
    return buf[0]<<8 |  buf[1];
}

float ADS1113_GetValue(void)
{
    UWORD value = 0;
    value = ADS1113_ReadData(); 
    float data = 0;
    if(value>>15){
        data = (((~value) & 0x7fff )* 2.048 / 0x7fff)*-1;
    }else {
        data = (((~value) & 0x7fff )* 2.048 / 0x7fff);
    }
    return data;
}




